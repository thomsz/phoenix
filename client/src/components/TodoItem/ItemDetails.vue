<template>
  <div class="p-4">
    <EditInPlace
      name="description"
      type="textarea"
      class="min-h-[2.5rem]"
      v-model="form.description"
    >
      <span v-if="form.description">{{ form.description }}</span>
      <span
        v-else
        class="text-slate-200"
      >
        Add description
      </span>
    </EditInPlace>
    <hr />
    <EditInPlace
      name="due_date"
      type="date"
      class="min-h-[2.5rem]"
      v-model="form.dueDate"
    >
      <span v-if="form.dueDate">{{ formattedLocaleDueDate }}</span>
      <span
        v-else
        class="text-slate-200"
      >
        Add due date
      </span>
    </EditInPlace>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import EditInPlace from '@/components/common/EditInPlace.vue'

import type Todo from '@/types/Todo'
import type { PropType } from 'vue'

export default defineComponent({
  name: 'ItemDetails',

  components: {
    EditInPlace
  },

  props: {
    modelValue: {
      type: Object as PropType<Todo>,
      required: true
    }
  },

  data () {
    return {
      form: this.modelValue
    }
  },

  computed: {
    formattedLocaleDueDate (): string | undefined {
      return new Date(this.form.dueDate).toLocaleDateString()
    }
  },

  watch: {
    form: {
      deep: true,
      handler (form: Todo) {
        this.$emit('update:modelValue', form)
      }
    }
  }
})
</script>
