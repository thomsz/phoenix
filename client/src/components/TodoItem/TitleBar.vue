<template>
  <div class="flex justify-between items-center h-12">
    <input
      name="task_done"
      type="checkbox"
      class="mx-2 h-full w-7 accent-green-600 hover:cursor-pointer"
      v-model="form.done"
    />
    <EditInPlace
      name="title"
      v-model="form.title"
    />
    <button
      class="action"
      data-testid="expand-button"
      @click="$emit('expand-button-clicked')"
    >
      <MenuAlt1Icon class="icon" />
    </button>
    <button
      class="action"
      data-testid="delete-item-button-clicked"
      @click="$emit('delete-item-button-clicked')"
    >
      <TrashIcon class="icon" />
    </button>
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import EditInPlace from '@/components/common/EditInPlace.vue'
import { TrashIcon, MenuAlt1Icon } from '@heroicons/vue/solid'

import type Todo from '@/types/Todo'
import type { PropType } from 'vue'

export default defineComponent({
  name: 'TitleBar',

  components: {
    TrashIcon,
    EditInPlace,
    MenuAlt1Icon
  },

  props: {
    modelValue: {
      type: Object as PropType<Todo>,
      required: true
    }
  },

  data () {
    return {
      form: this.modelValue
    }
  },

  watch: {
    form: {
      deep: true,
      handler (form: Todo) {
        this.$emit('update:modelValue', form)
      }
    }
  }
})
</script>

<style scoped>
button.action {
  @apply h-7 px-2 m-2 text-slate-300 hover:text-white hover:bg-slate-400 rounded transition-colors;
}
</style>
