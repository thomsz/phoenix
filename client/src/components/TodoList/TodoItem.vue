<template>
  <div>
    {{ item.title }}
    {{ item.description || '' }}
    {{ item.dueDate || '' }}
  </div>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import type { PropType } from 'vue'
import type Todo from '@/types/Todo'

export default defineComponent({
  name: 'TodoItem',

  props: {
    item: {
      type: Object as PropType<Todo>,
      required: true
    }
  }
})
</script>
