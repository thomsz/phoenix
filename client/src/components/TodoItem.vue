<template>
  <article
    class="hover:cursor-pointer hover:bg-slate-50 transition-colors duration-300"
    :class="{ 'opacity-25': item.done }"
  >
    <TitleBar
      v-model="form"
      @expand-button-clicked="expanded = !expanded"
      @delete-item-button-clicked="$emit('delete-item-button-clicked')"
    />
    <ItemDetails
      v-if="expanded"
      v-model="form"
    />
  </article>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

import TitleBar from '@/components/TodoItem/TitleBar.vue'
import ItemDetails from '@/components/TodoItem/ItemDetails.vue'

import type Todo from '@/types/Todo'
import type { PropType } from 'vue'

export default defineComponent({
  name: 'TodoItem',

  components: {
    TitleBar, 
    ItemDetails
  },

  props: {
    item: {
      type: Object as PropType<Todo>,
      required: true
    }
  },

  data () {
    const getFormattedDueDate = (): string | undefined => this.item.dueDate.split('T')[0]
    return {
      form: {
        ...this.item,
        dueDate: getFormattedDueDate() || ''
      },
      expanded: false
    }
  },

  watch: {
    form: {
      deep: true,
      handler (form: Todo) {
        const preparedPayload = {
          ...form,
          dueDate: form.dueDate ? new Date(form.dueDate).toISOString() : ''
        }

        this.$emit('item-changed', preparedPayload)
      }
    }
  }
})
</script>
