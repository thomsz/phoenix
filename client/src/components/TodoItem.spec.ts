import factory from '@/tests/services/factory'
import { faker } from '@faker-js/faker'
import { shallowMount } from '@vue/test-utils'

import TodoItem from '@/components/TodoItem.vue'
import TitleBar from '@/components/TodoItem/TitleBar.vue'
import ItemDetails from '@/components/TodoItem/ItemDetails.vue'

const mountingOptions = {
  props: {
    item: factory.generate('Todo')
  }
}

describe('TodoItem', () => {
  [
    ['done', true],
    ['undone', false]
  ].forEach(([descriptor, isDone]) => {
    it(`should render item done status: ${descriptor}`, () => {
      const wrapper = shallowMount(TodoItem, {
        ...mountingOptions,
        props: {
          ...mountingOptions.props,
          item: {
            ...mountingOptions.props.item,
            done: isDone
          }
        }
      })

      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('should emit event when clicking on delete item', () => {
    const wrapper = shallowMount(TodoItem, mountingOptions)
    const titleBarWrapper = wrapper.findComponent(TitleBar)
    titleBarWrapper.vm.$emit('delete-item-button-clicked')
    expect(wrapper.emitted('delete-item-button-clicked')).toBeTruthy()
  })

  it('should toggle details', async () => {
    const wrapper = shallowMount(TodoItem, mountingOptions)
    expect(wrapper.findComponent(ItemDetails).exists()).toBe(false)

    const titleBarWrapper = wrapper.findComponent(TitleBar)
    await titleBarWrapper.vm.$emit('expand-button-clicked')
    expect(wrapper.findComponent(ItemDetails).exists()).toBe(true)

    await titleBarWrapper.vm.$emit('expand-button-clicked')
    expect(wrapper.findComponent(ItemDetails).exists()).toBe(false)
  })

  it('should format due date', async () => {
    const wrapper = shallowMount(TodoItem, {
      ...mountingOptions,
      props: {
        ...mountingOptions.props,
        item: {
          ...mountingOptions.props.item,
          dueDate: '2022-08-20T10:15:59.765Z'
        }
      },
      data: () => ({
        expanded: true
      })
    })

    expect(wrapper.findComponent(ItemDetails).props('modelValue').dueDate).toBe('2022-08-20')
  })

  it('should emit event when item changes', async () => {
    const wrapper = shallowMount(TodoItem, mountingOptions)
    const updatedItem = {
      ...mountingOptions.props.item,
      title: faker.random.words()
    }

    await wrapper.findComponent(TitleBar).vm.$emit('update:modelValue', updatedItem)
    expect(wrapper.emitted('item-changed')?.[0][0]).toEqual({
      ...updatedItem,
      dueDate: new Date(updatedItem.dueDate).toISOString()
    })
  })
})
