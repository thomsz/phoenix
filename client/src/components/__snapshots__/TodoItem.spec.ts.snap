// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem should render item done status: done 1`] = `
"<article class=\\"hover:cursor-pointer hover:bg-slate-50 transition-colors duration-300 opacity-25\\">
  <title-bar-stub modelvalue=\\"[object Object]\\"></title-bar-stub>
  <!--v-if-->
</article>"
`;

exports[`TodoItem should render item done status: undone 1`] = `
"<article class=\\"hover:cursor-pointer hover:bg-slate-50 transition-colors duration-300\\">
  <title-bar-stub modelvalue=\\"[object Object]\\"></title-bar-stub>
  <!--v-if-->
</article>"
`;
